com.unity.netcode.gameobjects/Tests/Editor/Serialization/FastBufferWriterTests.cs

```diff
@@ -0,0 +1,2191 @@
+
+                WriteCheckBytes(ref writer, writeSize);
+
+                VerifyArrayEquality(valueToTest, writer.GetUnsafePtr(), 0);
+
+                var underlyingArray = writer.ToArray();
+                VerifyCheckBytes(underlyingArray, writeSize);
+            }
+        }
+
+        private unsafe void RunObjectTypeArrayTest<T>(T[] valueToTest) where T : unmanaged
+        {
+            var writeSize = FastBufferWriter.GetWriteSize(valueToTest);
+            // Extra byte for WriteObject adding isNull flag
+            var writer = new FastBufferWriter(writeSize + 3, Allocator.Temp);
+            using (writer)
+            {
+
+                Assert.AreEqual(sizeof(int) + sizeof(T) * valueToTest.Length, writeSize);
+
+                writer.WriteObject(valueToTest);
+                Assert.AreEqual(0, writer.ToArray()[0]);
+                VerifyPositionAndLength(ref writer, writeSize + sizeof(byte));
+
+                WriteCheckBytes(ref writer, writeSize + sizeof(byte));
+
+                VerifyArrayEquality(valueToTest, writer.GetUnsafePtr(), sizeof(byte));
+
+                var underlyingArray = writer.ToArray();
+                VerifyCheckBytes(underlyingArray, writeSize + sizeof(byte));
+            }
+        }
+        #endregion
+
+        #region Helpers
+        private TestStruct GetTestStruct()
+        {
+            var random = new Random();
+
+            var testStruct = new TestStruct
+            {
+                A = (byte)random.Next(),
+                B = (short)random.Next(),
+                C = (ushort)random.Next(),
+                D = (int)random.Next(),
+                E = (uint)random.Next(),
+                F = ((long)random.Next() << 32) + random.Next(),
+                G = ((ulong)random.Next() << 32) + (ulong)random.Next(),
+                H = true,
+                I = '\u263a',
+                J = (float)random.NextDouble(),
+                K = random.NextDouble(),
+            };
+
+            return testStruct;
+        }
+        #endregion
+
+        #region Tests
+        [Test]
+        public void TestWritingBasicTypes(
+            [Values(typeof(byte), typeof(sbyte), typeof(short), typeof(ushort), typeof(int), typeof(uint),
+                typeof(long), typeof(ulong), typeof(bool), typeof(char), typeof(float), typeof(double),
+                typeof(ByteEnum), typeof(SByteEnum), typeof(ShortEnum), typeof(UShortEnum), typeof(IntEnum),
+                typeof(UIntEnum), typeof(LongEnum), typeof(ULongEnum), typeof(Vector2), typeof(Vector3), typeof(Vector4),
+                typeof(Quaternion), typeof(Color), typeof(Color32), typeof(Ray), typeof(Ray2D))]
+            Type testType,
+            [Values] WriteType writeType)
+        {
+            var random = new Random();
+
+            if (testType == typeof(byte))
+            {
+                byte b = (byte)random.Next();
+                if (writeType == WriteType.WriteDirect)
```

Comment:
nit: You can probably save yoursel some 2-300 lines of repeats here by using a local function like

```
            void RunTypeTestLocal<T>(T val, WriteType wt) where T : unmanaged
            {
                switch (wt)
                {
                    case WriteType.WriteDirect:
                        RunTypeTest(val);
                        break;
                    case WriteType.WriteSafe:
                        RunTypeTestSafe(val);
                        break;
                    default:
                        RunObjectTypeTest(val);
                        break;
                }
            }
```